pkg/storage/factory.go

```diff
@@ -65,6 +71,98 @@ type NamedStores struct {
 	Swift      map[string]openstack.SwiftConfig     `yaml:"swift"`
 }
 
+func (ns NamedStores) checkForDuplicates() error {
+	t := reflect.TypeOf(ns)
+	v := reflect.ValueOf(ns)
+
+	seen := make(map[string]string)
+	// iterate over the struct fields
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		if field.Type.Kind() != reflect.Map {
+			continue
+		}
+
+		objectType := getYAMLTagKey(field)
+		// ignore fields without a valid yaml name
+		if objectType == "" {
+			continue
+		}
+
+		for _, key := range v.FieldByIndex(field.Index).MapKeys() {
+			if key.Kind() != reflect.String {
+				break
+			}
+
+			name := key.Interface().(string)
+
+			// return error if the name matches a predefined storage type
+			switch name {
+			case config.StorageTypeAWS, config.StorageTypeAWSDynamo, config.StorageTypeS3,
+				config.StorageTypeGCP, config.StorageTypeGCPColumnKey, config.StorageTypeBigTable, config.StorageTypeBigTableHashed, config.StorageTypeGCS,
+				config.StorageTypeAzure, config.StorageTypeBOS, config.StorageTypeSwift, config.StorageTypeCassandra,
+				config.StorageTypeFileSystem, config.StorageTypeInMemory, config.StorageTypeGrpc:
+				return fmt.Errorf("named store %s should not match with the name of a predefined storage type", name)
+			}
+
+			if objectType, ok := seen[name]; ok {
+				return fmt.Errorf("named store %s is already defined under %s", name, objectType)
+			}
+			seen[name] = objectType
+		}
+	}
+
+	return nil
+}
+
+func (ns NamedStores) lookupStoreType(name string) (string, bool) {
+	t := reflect.TypeOf(ns)
+	v := reflect.ValueOf(ns)
+
+	// iterate over the struct fields
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+		if field.Type.Kind() != reflect.Map {
+			continue
+		}
+
+		objectType := getYAMLTagKey(field)
+		// ignore fields without a valid yaml name
+		if objectType == "" {
+			continue
+		}
+
+		// perform lookup and check for non-zero value
+		if v := v.FieldByIndex(field.Index).MapIndex(reflect.ValueOf(name)); v.IsValid() {
+			return objectType, true
+		}
+	}
+
+	return "", false
+}
+
+func (ns NamedStores) validate() error {
+	for name, awsCfg := range ns.AWS {
```

Comment:
Howcome we're only validating these 3, out of interest?

Reply:
only 3 out of the supported named stores have a validation method

Reply:
Again, just a suggestion for the future:
It'd be cleaner to iterate over each config and see if it satisfies 

```golang
interface {
  Validate() error
}
```

than special-casing these 3 stores. It also has the benefit of expressing the intent more clearly (as opposed to this code which seems to only validate these 3 for some unknown reason).